Fix backdrop image ordering regression in EF Core migration

seeyabye opened the pull request:
## Issue

After migrating from SQLite direct queries to Entity Framework Core, backdrop images are no longer displayed in the correct order. Specifically, `fanart.jpg` (which should have priority) is often displayed after images from the `extrafanart/` folder.

This affects the display when "Select view" is set to "Thumb" or "Thumb Card", the backdrop image shown becomes unpredictable.

  ### Root Cause

  **In v10.10.7:**
  - Images were stored as a serialized string in a single database column
  - Order was preserved during serialization/deserialization

  **In current main (EF Core):**
  - Images are stored in the `BaseItemImageInfos` table without an explicit order column
  - When loading images, no `OrderBy` clause was specified
  - EF Core returns images in undefined order (typically database row order)
  - Simple alphabetical sorting doesn't work because `extrafanart/fanart1.jpg` sorts before `fanart.jpg` alphabetically

  ### Solution

  Implements priority-based sorting that matches the `LocalImageProvider` discovery order:

  1. `{mediaFileName}-fanart.jpg` (highest priority)
  2. `fanart.jpg` (not in extrafanart folder)
  3. `fanart-N.jpg` (numbered backdrops)
  4. `background.jpg` / `background-N.jpg`
  5. `art.jpg` / `art-N.jpg`
  6. `extrafanart/*` folder contents (alphabetical)
  7. `backdrop.jpg` / `backdropN.jpg`
  8. Other images (lowest priority)

  Within each priority group, images are sorted by numeric index (if present) and then alphabetically by path for stability.

  ### Changes

  - **File**: `Jellyfin.Server.Implementations/Item/BaseItemRepository.cs`
  - Added `GetImageOrderPriority()` helper method to determine priority based on filename patterns
  - Added `GetNumericImageIndex()` helper method to extract numeric indices from filenames
  - Updated image mapping to apply priority-based ordering

  ### Testing

  With a file structure like:
```
  ├── fanart.jpg
  ├── extrafanart/
  │   ├── fanart1.jpg
  │   ├── fanart2.jpg
  │   └── ...
  └── video.mp4
```

  **Before**: `extrafanart/fanart1.jpg` often displayed first
  **After**: `fanart.jpg` correctly displays first

JPVenson commented:
i really dont see this as a bugfix though.

AFAIK we never had an explicit order and it was just ordered in that particular way accidentally

JPVenson commented:
that order was never explicitly defined or should have been expected. As there was no implicit order before, i argue you should have exepected a random order before.

theguymadmax commented:
A user could define the image order through the webUI, which is now impossible ( #15105)

This is a bugfix to me.

seeyabye committed:
Fix backdrop image ordering regression after EF Core migration by implementing runtime sorting that matches LocalImageProvider discovery order

JPVenson commented:
@theguymadmax i cannot find any place in the webUI that allows explict ordering of images, can you specify?

Shadowghost reviewed (approved):
IMO this is a sensible solution to fix it

ferferga commented:
@Shadowghost Funnily enough this doesn't surprise me at all 😅😂. Still, why not change that and do it properly? It requires a migration and scanning the filesystem. All in all, probably it's faster a single roundtrip to the DB than 2 roundtrips (DB + filesystem).

This way it's also predicatble and we avoid further missteps in the future.

Shadowghost commented:
Don't get me wrong, I fully support handling it in the DB

seeyabye committed:
Add database-level SortOrder for backdrop images

- Add SortOrder column to BaseItemImageInfo with composite index
- Implement SwapImageSortOrder for user customization
- Update SaveImages to preserve user-swapped SortOrder
- Add PopulateImageSortOrder migration for existing images
- Order images by SortOrder when loading from database

JPVenson commented on Jellyfin.Server.Implementations/Item/BaseItemRepository.cs:
```diff
@@ -561,9 +566,127 @@ public void SaveImages(BaseItemDto item)
+
+        // Use ExecuteSqlRaw with schema-qualified table name (from trusted EF metadata) and parameterized values
+        // Table name is safe to concatenate as it comes from EF Core model, not user input
+        var sql = $@"
```
Raw SQL will not be accepted

JPVenson commented on Jellyfin.Server/Migrations/Routines/PopulateImageSortOrder.cs:
```diff
@@ -0,0 +1,239 @@
+/// </summary>
+#pragma warning disable CS0618 // Type or member is obsolete
+[JellyfinMigration("2025-10-23T02:30:00", nameof(PopulateImageSortOrder), "A7B3F8E9-4C2D-4E1A-9B5F-6D8E3F2A1C9B")]
+internal class PopulateImageSortOrder : IMigrationRoutine
```
Manipulates database -> Needs backup attribute 
Use `IAsyncMigrationRoutine` instead

JPVenson commented on Jellyfin.Server/Migrations/Routines/PopulateImageSortOrder.cs:
```diff
@@ -0,0 +1,239 @@
+        using var context = _dbProvider.CreateDbContext();
+
+        // Get all distinct ItemIds that have images
+        var itemIdsWithImages = context.BaseItemImageInfos
```
No need to materialise the list here just to have it as a parameter in the next query. Use the IQueryable, EFCore is perfectly capable of translating that

JPVenson commented on Jellyfin.Server/Migrations/Routines/PopulateImageSortOrder.cs:
```diff
@@ -0,0 +1,239 @@
+        int errorCount = 0;
+
+        // Load all base item paths into memory for fast lookup
+        var itemPathLookup = context.BaseItems
```
Loading all BaseItems into memory also pushes a lot of ram which isn't nessesary here.

JPVenson commented on Jellyfin.Server/Migrations/Routines/PopulateImageSortOrder.cs:
```diff
@@ -0,0 +1,239 @@
+
+        for (int batchStart = 0; batchStart < imagesByItem.Count; batchStart += batchSize)
+        {
+            var batch = imagesByItem.Skip(batchStart).Take(batchSize).ToList();
```
Use the PartitionHelper instead and join in the navigation properties instead

seeyabye committed:
Improve PopulateImageSortOrder migration with async and memory efficiency

nielsvanvelzen commented on Emby.Server.Implementations/Localization/Core/mr.json:
```diff
@@ -132,5 +132,10 @@
+    "TaskMoveTrickplayImagesDescription": "लायब्ररीच्या सेटिंग्जप्रमाणे आधीपासून अस्तित्वात असलेल्या ट्रिकप्ले फाइल्सचे स्थान बदलते.",
+    "TaskCleanCollectionsAndPlaylistsDescription": "जे संग्रह आणि प्लेलिस्ट आता अस्तित्वात नाहीत, त्यांमधील घटक हटवते.",
+    "CleanupUserDataTask": "वापरकर्ता डेटाची स्वच्छता प्रक्रिया",
+    "CleanupUserDataTaskDescription": "९० दिवसांहून अधिक काळ अनुपस्थित असलेल्या माध्यमांवरील सर्व वापरकर्ता माहिती (जसे पाहण्याची स्थिती, आवडी इ.) हटवते."
```
translations should never be added via pull requests, only via translate.jellyfin.org.

seeyabye replied:
@nielsvanvelzen thanks for the comments!
I am actually not sure why this is getting added. I made no changes to `mr.json` and from the commits, it's being automatically prepended to my fix... am I missing something?

nielsvanvelzen replied:
Looking at the PR commits you may have accidentally pulled in some commits to the master branch, probably when switching the target to the release branch. You can fix it by dropping the first two commits from your branch with an interactive rebase in git.

seeyabye committed:
Changed composite index to include ImageType

JPVenson commented on src/Jellyfin.Database/Jellyfin.Database.Providers.Sqlite/Migrations/20251022214259_AddSortOrderToBaseItemImageInfo.cs:
```diff
@@ -0,0 +1,47 @@
+                nullable: false,
+                defaultValue: 0);
+
+            migrationBuilder.CreateIndex(
```
you can just run smth like this inbetween the statement to pre-populate the SortOrder correctly:

```sql
WITH ImageInfos AS 
(
    SELECT Id, ROW_NUMBER() OVER (PARTITION BY ItemId, ImageType ORDER BY DateModified) OrderId FROM BaseItemImageInfos 
)

UPDATE BaseItemImageInfos 
    SET SortOrder = (SELECT OrderId FROM ImageInfos WHERE BaseItemImageInfos.Id = ImageInfos.Id)
```

seeyabye committed:
Prepoluate SortOrder before correction

seeyabye committed:
Changed SaveImage logic

seeyabye committed:
Fixed SortOrder not being populated on new scans

seeyabye committed:
Cleanup

seeyabye committed:
Fixed SortOrder for Web Uploads

seeyabye committed:
Fixed AddImage's ordering to ensure consistency

seeyabye committed:
Cleanup and fixed SortOrder not updating upon removal of backdrop images

theguymadmax commented:
@seeyabye. It's going to be too late to merge this on the release branch, as we are preparing for a 12.0 release.  Once #16062 is merged, please rebase and target the master branch. Thanks.

theguymadmax commented:
@seeyabye  the above-mentioned PR has been merged. Please rebase and target the master branch.
